File: app/src/main/java/com/huanchengfly/tieba/api/adapters/ForumFuzzyMatchAdapter.java
package com.huanchengfly.tieba.api.adapters;

import androidx.annotation.Nullable;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.huanchengfly.tieba.api.models.SearchForumBean;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ForumFuzzyMatchAdapter implements JsonDeserializer<List<SearchForumBean.ForumInfoBean>> {
    @Override
    public List<SearchForumBean.ForumInfoBean> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        List<SearchForumBean.ForumInfoBean> forumInfoBeans = new ArrayList<>();
        if (json.isJsonArray()) {
            JsonArray jsonArray = json.getAsJsonArray();
            for (JsonElement element : jsonArray) {
                if (element.isJsonObject()) {
                    forumInfoBeans.add(getForumInfoBean(element.getAsJsonObject()));
                }
            }
        } else if (json.isJsonObject()) {
            JsonObject jsonObject = json.getAsJsonObject();
            for (Map.Entry<String, JsonElement> elementEntry : jsonObject.entrySet()) {
                JsonElement jsonElement = elementEntry.getValue();
                if (jsonElement.isJsonObject()) {
                    forumInfoBeans.add(getForumInfoBean(jsonElement.getAsJsonObject()));
                }
            }
        }
        return forumInfoBeans;
    }

    @Nullable
    private String getNonNullString(JsonElement jsonElement) {
        return jsonElement.isJsonNull() ? null : jsonElement.getAsString();
    }

    private SearchForumBean.ForumInfoBean getForumInfoBean(JsonObject jsonObject) {
        return new SearchForumBean.ForumInfoBean()
                .setForumId(jsonObject.get("forum_id").getAsInt())
                .setForumName(getNonNullString(jsonObject.get("forum_name")))
                .setForumNameShow(getNonNullString(jsonObject.get("forum_name_show")))
                .setAvatar(getNonNullString(jsonObject.get("avatar")))
                .setPostNum(getNonNullString(jsonObject.get("post_num")))
                .setConcernNum(getNonNullString(jsonObject.get("concern_num")))
                .setHasConcerned(jsonObject.get("has_concerned").getAsInt());
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/api/adapters/ForumFuzzyMatchAdapter.kt
package com.huanchengfly.tieba.post.api.adapters

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.huanchengfly.tieba.post.api.models.SearchForumBean
import java.lang.reflect.Type

class ForumFuzzyMatchAdapter : JsonDeserializer<List<SearchForumBean.ForumInfoBean>> {
    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext,
    ): List<SearchForumBean.ForumInfoBean> {
        val forumInfoBeans: MutableList<SearchForumBean.ForumInfoBean> = ArrayList()
        if (json.isJsonArray) {
            val jsonArray = json.asJsonArray
            for (element in jsonArray) {
                if (element.isJsonObject) {
                    forumInfoBeans.add(getForumInfoBean(element.asJsonObject))
                }
            }
        } else if (json.isJsonObject) {
            val jsonObject = json.asJsonObject
            for ((_, jsonElement) in jsonObject.entrySet()) {
                if (jsonElement.isJsonObject) {
                    forumInfoBeans.add(getForumInfoBean(jsonElement.asJsonObject))
                }
            }
        }
        return forumInfoBeans
    }

    private fun getNonNullString(jsonElement: JsonElement): String? {
        return if (jsonElement.isJsonNull) null else jsonElement.asString
    }

    private fun getForumInfoBean(jsonObject: JsonObject): SearchForumBean.ForumInfoBean {
        return SearchForumBean.ForumInfoBean(
            forumId = jsonObject["forum_id"].asLong,
            forumName = getNonNullString(jsonObject["forum_name"]),
            forumNameShow = getNonNullString(jsonObject["forum_name_show"]),
            avatar = getNonNullString(jsonObject["avatar"]),
            postNum = getNonNullString(jsonObject["post_num"]) ?: "0",
            concernNum = getNonNullString(jsonObject["concern_num"]) ?: "0",
            hasConcerned = jsonObject["has_concerned"].asInt
        )
    }
}
